akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (C) 2009-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.actor.typed.scaladsl
+
+import akka.actor.typed.{ ActorRef, Props, TypedAkkaSpec }
+import akka.pattern.AskTimeoutException
+import akka.testkit.typed.TestKit
+import akka.testkit.typed.scaladsl.TestProbe
+import akka.util.Timeout
+import com.typesafe.config.ConfigFactory
+
+import scala.concurrent.duration._
+import scala.reflect.ClassTag
+import scala.util.{ Failure, Success }
+
+object ActorContextAskSpec {
+  val config = ConfigFactory.parseString(
+    """
+      ping-pong-dispatcher {
+        executor = thread-pool-executor
+        type = PinnedDispatcher
+      }
+      snitch-dispatcher {
+        executor = thread-pool-executor
+        type = PinnedDispatcher
+      }
+    """)
+}
+
+class ActorContextAskSpec extends TestKit(ActorContextAskSpec.config) with TypedAkkaSpec {
+
+  "The Scala DSL ActorContext" must {
+
+    // FIXME we should have a test over remoting as well in typed-cluster
+
+    "provide a safe ask" in {
+      case class Ping(sender: ActorRef[Pong])
+      case class Pong(selfName: String, threadName: String)
+
+      val pingPong = spawn(Actor.immutable[Ping] { (ctx, msg) ⇒
+        msg.sender ! Pong(ctx.self.path.name, Thread.currentThread().getName)
+        Actor.same
+      }, "ping-pong", Props.empty.withDispatcherFromConfig("ping-pong-dispatcher"))
+
+      val probe = TestProbe[AnyRef]()
+
+      val snitch = Actor.deferred[Pong] { (ctx) ⇒
+
+        // Timeout comes from TypedAkkaSpec
+
+        ctx.ask(pingPong, Ping) {
+          case Success(pong) ⇒ Pong(ctx.self.path.name + "1", Thread.currentThread().getName)
+          case Failure(ex)   ⇒ throw ex
```

Comment:
I think we should have the version without `Success` / `Failure` as the default. In the end, the `Success` / `Failure` behavior was particularly thought to be used with ask that returns a `Future`. 

Isn't the only way to fail by replying with `akka.actor.Status.Failure`? Is that even supported in typed so far?

Reply:
The most common reason and why I think we need it is for when the timeout hits.

Reply:
How about being able to provide a custom timeout message instead? It could have a default message defined that will lead to default ask timeout handling behavior.

Reply:
`Status.Failure` would only be for timeouts, I agree that it's better to define a separate timeout message.

What would default ask timeout handling be? Perhaps only warning log?

Reply:
one edge case that we also might have to handle is if asking an untyped actor and it replies with `Status.Failure`

Reply:
Isn't the problem then, that you casted the untyped actor to the wrong type? For untyped actors that can reply with `Status.Failure` there should be a way to define the return type manually to `Try[XYZ]`. There could e.g. be an extra overload / method `askOrFail` where the `Try[T]` needs to be handled manually.

Reply:
Isn't the reason why that (`a.a.Status.Failure`) worked with untyped kind of accidental, I mean, we respond with this type so therefore users could mimic that (I don't think there is any support in the typed ask impl for that right now). 
I think we should encourage users to have explicit protocols if their actors respond on failures.

I was thinking that you pretty much always explicitly want to act on a failed ask to another actor, and that you need context for that. In what scenario don't you care more than that it goes to the log?

Reply:
good point, if you don't care you would use `tell`

I think there could still be cases where you want to `ask` from a typed actor to an old existing actor, where you can't change the old actor. I think it's fine to map such failure to the same message as the timeout message.

Reply:
yes, that is expected since we use the untyped ask underneath, the Status.Failure will fail the Promise, and so on